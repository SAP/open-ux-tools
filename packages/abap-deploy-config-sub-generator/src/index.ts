export { getAbapQuestions } from './app/questions';
export { indexHtmlExists, showOverwriteQuestion } from './utils';
export { AbapDeployConfigOptions, DeployProjectType } from './app/types';
