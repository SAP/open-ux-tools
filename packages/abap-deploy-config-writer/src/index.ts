import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import cloneDeep from 'lodash/cloneDeep';
import { addPackageDevDependency, FileName, readUi5Yaml, updatePackageScript } from '@sap-ux/project-access';
import { getDeployConfig, updateBaseConfig } from './config';
import { addUi5Dependency, writeUi5RepositoryFiles } from './file';
import {
    BUILD_SCRIPT,
    DEPLOY_SCRIPT,
    RIMRAF,
    RIMRAF_VERSION,
    UI5_TASK_FLATTEN_LIB,
    UI5_TASK_FLATTEN_LIB_VERSION,
    UNDEPLOY_SCRIPT
} from './constants';
import type { DeployConfigOptions } from './types';
import type { Editor } from 'mem-fs-editor';
import type { AbapDeployConfig } from '@sap-ux/ui5-config';

/**
 * Writes the template to the memfs editor instance.
 *
 * @param basePath - the base path
 * @param abapDeployConfig - the abap deploy config
 * @param options - the deploy config options
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function generate(
    basePath: string,
    abapDeployConfig: AbapDeployConfig,
    options?: DeployConfigOptions,
    fs?: Editor
): Promise<Editor> {
    const abapConfig = cloneDeep(abapDeployConfig);

    if (!fs) {
        fs = create(createStorage());
    }

    // base config
    const baseConfigFile = options?.baseFile ?? FileName.Ui5Yaml;
    const baseConfig = await readUi5Yaml(basePath, baseConfigFile, fs);
    const baseConfigPath = join(basePath, baseConfigFile);
    updateBaseConfig(baseConfigPath, baseConfig, fs);

    // deploy config
    const deployConfigFile = options?.deployFile ?? FileName.UI5DeployYaml;
    const deployFilePath = join(basePath, deployConfigFile);
    const deployConfig = await getDeployConfig(abapConfig, baseConfig);
    fs.write(deployFilePath, deployConfig.toString());

    // package.json
    // deploy script
    const deployScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} && ${RIMRAF} archive.zip`;
    await updatePackageScript(basePath, 'deploy', deployScript, fs);
    // undeploy script
    const undeployScript = `${BUILD_SCRIPT} && ${UNDEPLOY_SCRIPT} --config ${deployConfigFile}`;
    await updatePackageScript(basePath, 'undeploy', undeployScript, fs);
    // test mode script
    const deployTestModeScript = `${BUILD_SCRIPT} && ${DEPLOY_SCRIPT} --config ${deployConfigFile} --testMode true`;
    await updatePackageScript(basePath, 'deploy-test', deployTestModeScript, fs);
    // dependencies
    await addPackageDevDependency(basePath, RIMRAF, RIMRAF_VERSION, fs);
    await addPackageDevDependency(basePath, UI5_TASK_FLATTEN_LIB, UI5_TASK_FLATTEN_LIB_VERSION, fs);
    addUi5Dependency(fs, basePath, UI5_TASK_FLATTEN_LIB);

    // ui5 repo files
    await writeUi5RepositoryFiles(fs, basePath);

    return fs;
}

export { generate };
