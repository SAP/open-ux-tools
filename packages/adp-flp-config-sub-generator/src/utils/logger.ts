import { DefaultLogger, LogWrapper, type ILogWrapper } from '@sap-ux/fiori-generator-shared';
import type { Logger } from 'yeoman-environment';
import type { IVSCodeExtLogger, LogLevel } from '@vscode-logging/logger';

/**
 * Static logger prevents passing of logger references through all functions, as this is a cross-cutting concern.
 */
export default class AdpFlpConfigLogger {
    private static _logger: ILogWrapper = DefaultLogger;

    /**
     * Get the logger.
     *
     * @returns the logger
     */
    public static get logger(): ILogWrapper {
        return AdpFlpConfigLogger._logger;
    }

    /**
     * Set the logger.
     *
     * @param value the logger to set
     */
    public static set logger(value: ILogWrapper) {
        AdpFlpConfigLogger._logger = value;
    }

    /**
     * Configures the vscode logger.
     *
     * @param vscLogger - the vscode logger
     * @param loggerName - the logger name
     * @param yoLogger - the yeoman logger
     * @param vscode - the vscode instance
     * @param logLevel - the log level
     */
    static configureLogging(
        vscLogger: IVSCodeExtLogger,
        loggerName: string,
        yoLogger: Logger,
        vscode?: unknown,
        logLevel?: LogLevel
    ): void {
        const logWrapper = new LogWrapper(loggerName, yoLogger, logLevel, vscLogger, vscode);
        AdpFlpConfigLogger.logger = logWrapper;
    }
}
