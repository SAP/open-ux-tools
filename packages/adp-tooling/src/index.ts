export { AdpPreviewConfig, AdpWriterConfig } from './types';
export * from './preview/adp-preview';
export { generate } from './writer';
export * from './base/prompt';
