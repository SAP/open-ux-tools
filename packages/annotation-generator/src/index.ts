export { AnnotationServiceParameters, GenerateAnnotationsOptions, generateAnnotations } from './generation';
