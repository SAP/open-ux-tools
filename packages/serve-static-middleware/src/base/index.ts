export { serveStaticMiddleware } from './servestatic';
export { ServeStaticConfig } from './types';
