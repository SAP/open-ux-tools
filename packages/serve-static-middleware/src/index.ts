export { serveStaticMiddleware, ServeStaticConfig } from './base';
