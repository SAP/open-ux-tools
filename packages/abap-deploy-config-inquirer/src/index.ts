import { ToolsLogger, type Logger } from '@sap-ux/logger';
import { initI18n } from './i18n';
import { PromptState } from './prompts/prompt-state';
import { getAbapDeployConfigQuestions } from './prompts';
import LoggerHelper from './logger-helper';
import type { InquirerAdapter } from '@sap-ux/inquirer-common';
import type { AbapDeployConfigAnswers, AbapDeployConfigPromptOptions } from './types';

/**
 * Get the inquirer prompts for abap deploy config.
 *
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompts used to provide input for abap deploy config generation and a reference to the answers object which will be populated with the user's responses once `inquirer.prompt` returns
 */
async function getPrompts(
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<{ prompts: any[]; answers: Partial<AbapDeployConfigAnswers> }> {
    await initI18n();
    LoggerHelper.logger = logger ?? new ToolsLogger({ logPrefix: '@sap-ux/abap-deploy-config-inquirer' });

    return {
        prompts: await getAbapDeployConfigQuestions(promptOptions),
        // Return reference to derived answers object that will be populated with user responses (after prompting is complete)
        answers: PromptState.abapDeployConfig
    };
}

/**
 * Prompt for abap deploy config writer inputs.
 *
 * @param adapter - optionally provide references to a calling inquirer instance, this supports integration to Yeoman generators, for example
 * @param promptOptions - options that can control some of the prompt behavior. See {@link AbapDeployConfigPromptOptions} for details
 * @param logger - a logger compatible with the {@link Logger} interface
 * @returns the prompt answers
 */
async function prompt(
    adapter: InquirerAdapter,
    promptOptions?: AbapDeployConfigPromptOptions,
    logger?: Logger
): Promise<AbapDeployConfigAnswers> {
    const abapDeployConfigPrompts = (await getPrompts(promptOptions, logger)).prompts;
    const answers = await adapter.prompt<AbapDeployConfigAnswers>(abapDeployConfigPrompts);

    // Add dervied service answers to the answers object
    Object.assign(answers, PromptState.abapDeployConfig);

    return answers;
}

export { getPrompts, prompt, type InquirerAdapter, type AbapDeployConfigAnswers, type AbapDeployConfigPromptOptions };
