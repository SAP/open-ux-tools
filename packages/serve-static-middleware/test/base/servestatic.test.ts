import { serveStaticMiddleware } from '../../src';
import type { ServeStaticConfig } from '../../src';
import { NullTransport, ToolsLogger } from '@sap-ux/logger';
import * as expressServeStatic from 'serve-static';
import { relative, join } from 'path';

jest.mock('serve-static', () => ({
    __esModule: true,
    default: jest.fn()
}));

describe('serve-static-middleware', () => {
    const logger = new ToolsLogger({
        transports: [new NullTransport()]
    });
    const serveStaticMock = jest.spyOn(expressServeStatic as any, 'default').mockImplementation(() => {
        return (req: any, res: any, next: any) => {
            next();
        };
    });
    const configuration: ServeStaticConfig = {
        paths: [
            { path: '/resources', src: '/path/to/resources' },
            { path: '/test-resources', src: '/path/to/test-resources' }
        ]
    };

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('serveStaticMiddleware: call with minimal config', () => {
        serveStaticMiddleware('/root', configuration, logger);
        expect(serveStaticMock).toHaveBeenCalledTimes(2);
        expect(serveStaticMock.mock.calls).toMatchSnapshot();
    });

    test('serveStaticMiddleware: call with additional options', () => {
        const config: ServeStaticConfig = {
            paths: [
                { path: '/resources', src: '/path/to/resources', maxAge: 123, fallthrough: false },
                { path: '/test-resources', src: '/path/to/test-resources', maxAge: 456, fallthrough: true }
            ]
        };
        serveStaticMiddleware('/root', config, logger);
        expect(serveStaticMock).toHaveBeenCalledTimes(2);
        expect(serveStaticMock.mock.calls).toMatchSnapshot();
    });

    test('serveStaticMiddleware: call with src path resolution', () => {
        const config: ServeStaticConfig = {
            paths: [
                { path: '/resources', src: 'resources' },
                { path: '/test-resources', src: 'test-resources' }
            ]
        };
        serveStaticMiddleware(__dirname, config, logger);
        expect(serveStaticMock).toHaveBeenCalledTimes(2);
        expect(serveStaticMock).toHaveBeenNthCalledWith(1, relative(process.cwd(), join(__dirname, 'resources')), {});
        expect(serveStaticMock).toHaveBeenNthCalledWith(
            2,
            relative(process.cwd(), join(__dirname, 'test-resources')),
            {}
        );
    });

    test('serveStaticMiddleware: call with global options', () => {
        const config: ServeStaticConfig = {
            dotfiles: 'ignore',
            etag: false,
            fallthrough: false,
            paths: [
                { path: '/resources', src: '/path/to/resources' },
                { path: '/test-resources', src: '/path/to/test-resources' }
            ]
        };
        serveStaticMiddleware('/root', config, logger);
        expect(serveStaticMock).toHaveBeenCalledTimes(2);
        expect(serveStaticMock.mock.calls).toMatchSnapshot();
    });
});
