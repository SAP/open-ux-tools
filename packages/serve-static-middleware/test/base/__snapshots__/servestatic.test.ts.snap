// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serve-static-middleware serveStaticMiddleware: call with additional options 1`] = `
Array [
  Array [
    "/path/to/resources",
    Object {
      "fallthrough": false,
      "maxAge": 123,
    },
  ],
  Array [
    "/path/to/test-resources",
    Object {
      "fallthrough": true,
      "maxAge": 456,
    },
  ],
]
`;

exports[`serve-static-middleware serveStaticMiddleware: call with global options 1`] = `
Array [
  Array [
    "/path/to/resources",
    Object {
      "dotfiles": "ignore",
      "etag": false,
      "fallthrough": false,
    },
  ],
  Array [
    "/path/to/test-resources",
    Object {
      "dotfiles": "ignore",
      "etag": false,
      "fallthrough": false,
    },
  ],
]
`;

exports[`serve-static-middleware serveStaticMiddleware: call with minimal config 1`] = `
Array [
  Array [
    "/path/to/resources",
    Object {},
  ],
  Array [
    "/path/to/test-resources",
    Object {},
  ],
]
`;
